v2/rust/src/child_process_supervisor.rs

```diff
@@ -0,0 +1,93 @@
+use super::termination::TerminationFlag;
+use super::timeout::Timeout;
+use anyhow::{bail, Context, Result};
+use log::{error, warn};
+use std::collections::{HashMap, HashSet};
+use std::process::{Child, Command, ExitStatus};
+use std::thread::sleep;
+use std::time::Duration;
+use sysinfo::{Pid, PidExt, Process, ProcessExt, System, SystemExt};
+
+pub struct ChildProcessSupervisor<'a> {
+    pub command: Command,
+    pub timeout: u64,
+    pub termination_flag: &'a TerminationFlag,
+}
+
+impl ChildProcessSupervisor<'_> {
+    pub fn run(mut self) -> Result<ChildProcessOutcome> {
+        let mut child = self.command.spawn().context("Failed to spawn subprocess")?;
+        let timeout = Timeout::start(self.timeout);
+
+        loop {
+            if let Some(exit_status) = child
+                .try_wait()
+                .context(format!(
+                    "Failed to query exit status of process {}, killing",
+                    child.id()
+                ))
+                .map_err(|err| {
+                    kill_child_process_tree(&mut child);
+                    err
+                })?
+            {
+                return Ok(ChildProcessOutcome::Exited(exit_status));
+            }
+
+            if timeout.expired() {
+                error!("Process timed out");
+                kill_child_process_tree(&mut child);
+                return Ok(ChildProcessOutcome::TimedOut);
+            }
+
+            if self.termination_flag.should_terminate() {
+                warn!("Terminated");
+                kill_child_process_tree(&mut child);
+                bail!("Terminated")
+            }
+            sleep(Duration::from_millis(250))
+        }
+    }
+}
+
+pub enum ChildProcessOutcome {
+    Exited(ExitStatus),
+    TimedOut,
+}
+
+fn kill_child_process_tree(child: &mut Child) {
```

Comment:
`kill_child_and_process_tree` and `kill_all_children` feel like the same function. Consider combining them into one function.

Reply:
Or maybe we could rename `kill_child_process_tree(child: &mut Child)` to `kill_process_tree(root_process: &mut Child)`?

Reply:
Done